Improve mechanism for wrapping errors e.g. createContainerError

markfields opened the pull request:
We have code that takes an arbitrary `error` and converts it to a more canonical error object with a meaningful type and also capabilities like safe logging of props.  This was implemented a few different places with slightly different semantics - and some bugs (e.g. the pitfall of #6468). This PR unifies this to a single `wrapError` function that takes a callback allowing for wrapping in different types of errors, but sharing the core logic / semantics.

markfields committed:
Refactoring with two small behavior changes.

Manually porting from old markfields:container-error branch

markfields committed:
Fix build failures

markfields committed:
Some comments/explorations

markfields committed:
lock

markfields committed:
make LoggingError abstract and implement IErrorBase.

And it builds!

markfields committed:
Did strict null checks turn on suddenly?

markfields committed:
Introduce wrapError helper and add a few new usages

markfields committed:
Simpler wrapping semantics

CreateContainerError (and CreateProcessingError)
do their own thing before calling wrapError

markfields committed:
Simpler/clearer wrapObject function

markfields committed:
Merge remote-tracking branch 'origin/main' into errors-cleanup

markfields commented:
/azp build-client

vladsud commented on common/lib/driver-definitions/src/driverError.ts:
```diff
@@ -109,7 +109,7 @@ export interface IAuthorizationError extends IDriverErrorBase {
 export interface IDriverBasicError extends IDriverErrorBase {
     readonly errorType:
-    DriverErrorType.genericError
+    DriverErrorType.genericError  //* seems unused
```
We should delete if unused.

markfields replied:
I'll consider in a follow-up PR

vladsud commented on packages/loader/container-loader/src/deltaManager.ts:
```diff
@@ -732,8 +734,8 @@ export class DeltaManager
-            const error = new LoggingError("Op is sent in read-only document state", {
-                errorType: ContainerErrorType.genericError,
+            const error = new GenericError("Op is sent in read-only document state", {
+                errorType: ContainerErrorType.genericError,  //* use GenericError ctor?
```
May I suggest using PR comments instead of putting them in code? If the intention is not to merge them.

vladsud commented on packages/loader/container-utils/src/error.ts:
```diff
@@ -9,21 +9,21 @@ import {
 
 const isValidLoggingError = (error: any): error is LoggingError => {
-    return typeof error?.errorType === "string" &&  LoggingError.is(error);
+    return LoggingError.is(error);
```
Curious, if there is convention here. For example, Array.isArray

vladsud commented:
Can you please add at least some description? I.e. what exactly is the problem, what is the solution, etc.

vladsud commented on packages/utils/telemetry-utils/src/logger.ts:
```diff
@@ -577,6 +583,16 @@ export class LoggingError extends Error implements ILoggingError {
     }
 }
 
+export class SomeLoggingError extends LoggingError {
```
I do not know what "Some" means in this context.

markfields replied:
This was a placeholder name I forgot to fix up.  I need to ponder your comment above about making LoggingError abstract.  If I revert that part, then this class is no longer needed.  Might be unrelated to the main point of the PR and can be reverted anyway to be considered separately.

markfields replied:
Yep, reverting this.

vladsud commented on packages/utils/telemetry-utils/src/logger.ts:
```diff
@@ -533,14 +536,17 @@ function getValidTelemetryProps(obj: any): ITelemetryProperties {
+export abstract class LoggingError extends Error implements ILoggingError, Omit<IErrorBase, "sequenceNumber"> {
     private readonly __isFluidLoggingError__ = 1;
 
+    public abstract readonly errorType: string;
```
I'm not sure about errorType here (and abstract class).
Basically the main purpose of errorType, or rather where it is used - communicate to host about what error happens on Container.load() and Container's "closed" event paths. We lot same thing to telemetry, which is fine, but it's not the primary reason for its existence.

LoggingError on the other hand should likely be viewed as an error object that helps us track what is safe and not save to be logged to telemetry. In general, it does not have to conform to requirements above. And vice versa, the error we log to host does not have to have anything safe to be logged to telemetry. While it's true that likely 99% of usage would overlap here, I think I'd prefer logging layer (this layer) have zero knowledge of above. If needed, we can add another derived class on top of it somewhere else (container-utils?) that does rely on both concepts.

markfields replied:
Yes I got more clarity on the separate abstractions of errorType/interfaces v. LoggingError as I worked on the PR, and reading what you're saying here I agree on reverting this.

markfields replied:
Another hint is the policy check failure which is about a layer-check violation :)

markfields committed:
Fix existing UTs

markfields committed:
Remove //* To-do comments

markfields committed:
Remove errorType from loggingError

Also stop overwriting GenericError's errorType
if found on wrapped error.
Log with different property key instead.

markfields committed:
fix build break

markfields commented:
@vladsud - Ready for you to look at for real now.  Thanks!

markfields committed:
Fix test failures

markfields commented on packages/drivers/driver-base/src/documentDeltaConnection.ts:
```diff
@@ -419,7 +419,9 @@ export class DocumentDeltaConnection
 
     private removeConnectionListeners() {
-        clearTimeout(this.socketConnectionTimeout);
+        if (this.socketConnectionTimeout !== undefined) {
```
There were a few random build errors, as if strict null checks was just turned on or something... Can't hurt I guess.

markfields commented on packages/loader/driver-utils/src/multiDocumentServiceFactory.ts:
```diff
@@ -45,7 +45,8 @@ export class MultiDocumentServiceFactory implements IDocumentServiceFactory {
         ensureFluidResolvedUrl(resolvedUrl);
         const urlObj = parse(resolvedUrl.url);
-        if (urlObj.protocol === undefined) {
+        // eslint-disable-next-line no-null/no-null
+        if (urlObj.protocol === undefined || urlObj.protocol === null) {
```
Another random stict null checks fix - not sure why these started breaking.

markfields committed:
Polish

markfields commented:
@jcmoore - no pressure since you're OCE but this would be an interesting PR to you.

markfields committed:
Build/Lint errors

vladsud commented on packages/loader/container-utils/src/error.ts:
```diff
@@ -3,27 +3,26 @@
 
-const isValidLoggingError = (error: any): error is LoggingError => {
-    return typeof error?.errorType === "string" &&  LoggingError.is(error);
+const isValidLoggingError = (error: any): error is LoggingError & IErrorBase => {
```
Since you are poking at it, can you please consider giving it different name to better differentiate from 
isILoggingError?
"valid" is useless (I think it should be removed), and the rest looks almost identical to the other one, making it hard to see the difference.

markfields replied:
Yeah I've been pondering about when we should be pivoting on an object being a concrete `LoggingError` instance (as best we can) v. adhering to `ILoggingError`. I wrote `isILoggingError` and then @anthony-murphy wrote `LoggingError.is`.

In this case, "valid" I think refers to the fact that it also has `errorType`.  I will take a moment to think over a better name/factoring.

markfields replied:
Added a comment. it needs to stay as-is

vladsud commented on packages/loader/container-utils/src/error.ts:
```diff
@@ -85,9 +110,38 @@ export class DataProcessingError extends LoggingError implements IErrorBase {
+            (errMsg: string, props?: ITelemetryProperties) =>
+                new DataProcessingError(errMsg, { ...props, ...messagePropsToLog });
+
+        return wrapError(error, newErrorFn);
```
I think the following flow is a bit more readable and ensures that properties always applied in the same order in both branches:

```typescript
const error2 = isValidLoggingError(error) ? error : wrapError(error, DataProcessingError);
error2.addTelemetryProperties(messagePropsToLog);
return error2;
```

Assumes 
```
export function wrapError<T extends LoggingError>(
    error: any,
    newErrorFn: new (m: string, p?: ITelemetryProperties) => T,
```

i.e. that ctor is passed in for an object.

vladsud replied:
Similar feedback for other places (like CreateContainerError)

markfields replied:
Great feedback.  I can't do the `newErrorFn: new`... part because sometimes you need a more interesting function.  But otherwise, did it.

anthony-murphy commented on packages/loader/container-utils/src/error.ts:
```diff
@@ -3,27 +3,26 @@
+        return error.message as string;
     }
-    return `${error}`;
+    return String(error);
```
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#number-string-boolean-symbol-and-object

anthony-murphy replied:
maybe this usage is ok, but i prefer the old way and not to introduce any upper case String usages

markfields replied:
The old way throws an exception if `error` is a `Symbol`. This way was recommended on SO or the TS docs or something.

anthony-murphy commented on packages/loader/container-utils/src/error.ts:
```diff
@@ -55,17 +54,43 @@ function extractLogSafeErrorProperties(error: any) {
-        errorMessage: string,
-        readonly error: any,
+        message: string,
         props?: ITelemetryProperties,
```
this change is risky, as i don't think you'll get a compile time error for any missed cases

markfields replied:
Hmmmmm good point, I do remember getting test failures for it.  I'll double-check usages. I don't imagine anyone outside our codebase is using this constructor either...

markfields replied:
Reverted it, thanks for pointing this out.

markfields committed:
PR feedback

anthony-murphy commented on packages/loader/container-utils/src/error.ts:
```diff
@@ -85,9 +111,37 @@ export class DataProcessingError extends LoggingError implements IErrorBase {
+                new DataProcessingError(errMsg);
+
+        // Don't coerce if it's already a recognized LoggingError
+        const error = isValidLoggingError(originalError)
```
this code is repeated here and in CreateContainerError why not just have wrapError return the input error if isValidLoggingError, or am i missing cases where we wrap an loggingError

anthony-murphy replied:
i don't see where it is used, but i see how it works in wrap error now

markfields replied:
Yes there are other cases where we wrap a loggingError.  See ThrottlingWarning and SummarizerError or whatever they're called.

markfields committed:
PR Feedback and API docs update

markfields committed:
Test fix

markfields committed:
API docs

markfields commented:
Just found another bug this fixes - `CreateProcessingError` used to not log all the message metadata if the incoming error was already a "valid LoggingError" - which is often the case (the driver emits "nice" errors like this)
